File: greedy/1774_closest_dessert_cost/1774_closest_dessert_cost.py
from typing import List

class Solution:
    def closestCost(self, baseCosts: List[int], toppingCosts: List[int], target: int) -> int:
        ans = baseCosts[0]
        li = baseCosts
        for cost in toppingCosts:
            for c in li.copy():
                li.append(c + cost)
                li.append(c + 2 * cost)
        for a in li:
            if abs(a - target) < abs(ans - target) or abs(a - target) == abs(ans - target) and a < ans:
                ans = a
        return ans



baseCosts = [1,7]
toppingCosts = [3,4]
target = 10
res = Solution().closestCost(baseCosts, toppingCosts, target)
print(res)

File: greedy/1774_closest_dessert_cost/1774_closest_dessert_cost.go
package main 

import "fmt"
import "math"


func closestCost(baseCosts []int, toppingCosts []int, target int) int {
	res := math.MinInt32
	var dfs func(idx int, need int) 
	dfs = func(idx int, need int) {
		if idx == len(toppingCosts) {
			if abs(res - target) > abs(need - target) {
				res = need
			} else if abs(res - target) == abs(need-target) {
				res = min(res, need)
			}
			return
		}
		dfs(idx + 1, need)
		dfs(idx + 1, need + toppingCosts[idx])
		dfs(idx + 1, need + toppingCosts[idx] * 2)
	}

	for i := 0; i < len(baseCosts); i++ {
		dfs(0, baseCosts[i])
	}
	return res
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	baseCosts := []int{1,7}
	toppingCosts := []int{3,4}
	target := 10
	res := closestCost(baseCosts, toppingCosts, target)
	fmt.Println(res)
}